### ISSUE TITLE
[feat] Vertical Mapped Stochastic Simulation of Observables under noise (Quantum Trajectories)

### ISSUE THREAD

**Issue by onurdanaci** [qulacs/qulacs#530]
Hi,

As in Cirq, Tensorcircuit and some other quantum circuit simulators, Qulacs is supporting [1] the Quantum Trajectories approach while simulating observable expectations after quantum channels. In other words, in the context of StateVector simulation instead of DensityMatrix simulation, Qulacs allow you to define unary channels (such as Depolarizing, Dephasing, etc.) as probabilistic unitary gates such that when you compute a state vector, you sample a different vector each time. You can use each of these samples as Monte Carlo trajectories, get an expectation value out of them, then take their mean / average at the end. This is an efficient approach, albeit noisy one, comparing to slow yet deterministic DensityMatrix simulation.

In the new state simulator Tensorcircuit, you are allowed to perform this computation even more efficiently by "externalizing randomness" [2]. In other words, you are allowed to sample all the random numbers offline (ahead of the noisy simulation) so that your quantum channels, which are also defined as random unitary gates in Tensorcircuit, can use these pre-computed random numbers instead of computing random numbers online at each Monte Carlo trajectory. By doing so, it can use modern vectorization method of Vertical Mapping (VMAP) in machine learning frameworks (e.g, Jax, Pytorch, Tensorflow, etc.), and compute all of the noisy observable expectations at the same time, as an array. It also has the further advantage of making each of these runs 3x orders of magnitude faster by applying Just In Time Compilation (JIT) on them. Then, taking a mean of these expectations is already highly optimized, without requiring an explicit for loop. 

While performing an optimization subroutine to a parameteric quantum circuit under noise, it's not terribly important to have new randomly sampled numbers at each optimization iteration. e.g, If you pre-sampled 1000 random numbers out of uniform distribution for 1000 Monte Carlo trajectories ( nmc = 1000 ), it is more than OK to recycle them at each noisy optimization iteration for different parameters on your parametric quantum circuit. If you are taking the VMAP & JIT route, you only have one slow iteration at the beginning due to JIT compilation of your noisy expectation computation, then it is amortized by the fast evaluations further. This approach in its current application, on the other hand, has its own short-comings stemming from the rigidity of the JAX (or other) backend the simulator is using, as these backends are developed for other industrial applications by tech giants. For example, if you need to dynamically build parameterized quantum circuits out of a tensor/graph representation of a circuit by accessing circuit information from that tensor, this JIT won't work. A very simple example to give could be that let's say you have a 4 x 4 matrix M representing a four qubit circuit, and if its 0th and 1st element is one, M[0,1] = 1, then you are applying a X gate (the 0th pauli gate) on qubit 2 (1st element of 4 qubits). In JIT based simulators, accessing such dynamic circuit structure will throw an error.

In Qulacs, it is possible to pre-define a parametric quantum circuit PQC (e.g, based on circuit information coming from a matrix, or other), then feed it to a (stochastic) noisy observable simulation subroutine. The issue is, this simulation needs to be done very slowly in a Python for loop in range( nmc ), and at each Monte Carlo trajectory Qulacs unnecessarily spends some time computing new, original random numbers. If each state vector simulation takes "A" amount of wall-clock time, the whole computation takes at least "A x nmc" amount of wall-clock time. That is very inefficient. And, unlike the other simulators, it is very hard for an end user to get access to Qulacs' highly-optimized backend, as it is written C++, and it's opaque.

My feature request is to have random unitary gates, such as depolarizing channels having the option of externally feeding random numbers instead of computing each time, such as qulacs.gates.depolarizingchannels( p = 0.01, status = random_numbers_you_feed). And, once you had that, an efficient subroutine to perform Monte Carlo quantum trajectories based simulation as follows. Doing something in the lines of:

`
nmc = 1000
PQC = "your parametrized quantum circuit that was defined somewhere, somehow, it has random unitaries"
n_channels = get_number_of_randomchannels( PQC ) 
randoms_matrix = np.random.uniform( size = (nmc, n_channels) )

def noisy_circuit_evaluation(circuit, randoms, observable):
    #TODO : return expectation value of observable by feeding the random number on random unitary

def noisy_MC_evaluation( circuit, randoms_matrix, observable, nmc ):
    noisy_expect = 0 
    for imc in range( nmc ):
        randoms = random_matrix[imc, :]
        noisy_expect += noisy_circuit_evaluation(circuit, randoms, observable)

    return noisy_expect / nmc
`

Or performing these Vertical Mappings using new, auto-vectorization supporting C++ compilers so that user gets the averaged expectation after computing "nmc" many of them.

But defining these loops and everythin in an efficient, time optimized manner in C++ backend. So when we are using Qulacs' Python interface, our subroutine doesn't need to compute new random variables online at each Monte Carlo trajectory, or we don't need to use Python's extremely slow for-loop. 

Thank you for your time.
Best,
Onur

[1] http://docs.qulacs.org/en/latest/guide/2.0_python_advanced.html

[2]https://tensorcircuit.readthedocs.io/en/latest/whitepaper/5-density-matrix.html#Externalizing-the-Randomness

**Comment by forest1040** [comment]
Thank you.
The idea of parallelizing circuits by pre-sampling random numbers is very interesting.
I'll think about it.

**Comment by onurdanaci** [comment]
@forest1040 Do you know when would it be implemented? I'm going to schedule a project based on that information.

**Comment by forest1040** [comment]
@onurdanaci If so, please implement the feature there and send us a PR.

**Comment by onurdanaci** [comment]
@forest1040 I will implement this feature in the upcoming months.
